### PR TITLE
Gas Turbine

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's implementation of the gas turbine.

Tested and works--pretty neat little machine; can produce a surprising amount power (in an upgraded state) for just burning plasma.

Can be upgraded by R&D.

Added in the turbine to the map in the turbine room: https://github.com/ParadiseSS13/Paradise/pull/2153

### CODE REVIEW

**Comment by Robertha89** [comment]
According to this it generates 100k per rpm, and some of the numbers used are a bit screwy, confriction is labled as  #define COMPFRICTION 5e5, and thats 3125, the efficency = E / 6, where E is the rating of the manipulator component, with 6 manipulators, efficency is 1 with base manipulators, and with femto-manipulators it's 4

rpm = max(0, rpm - (rpm*rpm)/(COMPFRICTION/efficiency)
not sure how this is even calculated but say RPM is 50k, due to it having an icon state for such

50k - (50k*50K) / (3125/4) = -3.150.000
50k - 2.500.000.000 / 781.25 = -3.150.000
and this value does not seem correct, so i am assuming i am calculating it wrong somehow where the - does not mean minus but from to, with a max rpm of 3.200.000, assuming you reach that number, the max rpm will increase even more.

Anyway, there is a var called rpmtarget, which has an value of 0, which looks to have a max value of 1k and rpm is calculated by 0.9* rpm + 0.1 * rpmtarget, which against confuses me since if the rpm is 50k, it would be calculated to be 45100 here, but that might just be how fast the rpm drops if it aint gaining enough burning plasma... Either way, I am miss understanding the math, could you explain it for me? 

Cause of max rpm is 3.200.000 and you reach that, you would produce more then 160k times what the singularity produces, which seems wrong... on the other hand, if this is to be true, it would be amusing to see people getting shocked.

**Comment by Taaketa** [comment]
I'm excited by this. You are saying that you're adding this to the maint area in the other update? Pre-assembled? I take it sorta works like the TEG?

**Comment by taukausanake** [comment]
Can I throw shit into it and watch them melt/gib/explode/make the turbine explode? That's all I want to see from this. Screw powering the station

**Comment by Fox-McCloud** [comment]
Once that map change gets merged, then I'll update this PR to add it directly to maintenance.

Yes, it will start pre-assembled.

**Comment by Allfd** [comment]
Very excited!  

If only we had substations.

**Comment by Fox-McCloud** [comment]
Oh, it functions very differently from the TEG though. The TEG is largely powered by pressure and the difference between extremely cold temps and extremely high temps.

The gas turbine operates by pressure and high temperature.

**Comment by Fox-McCloud** [comment]
Mapped in the turbine.

File: code/modules/power/turbine.dm [hunk 1]
```diff
@@ -21,57 +47,116 @@
...
-	if(stat & BROKEN)
+	turbine = locate() in get_step(src, get_dir(inturf, src))
+	if(turbine)
+		turbine.locate_machinery()
+
+/obj/machinery/power/compressor/RefreshParts()
+	var/E = 0
+	for(var/obj/item/weapon/stock_parts/manipulator/M in component_parts)
+		E += M.rating
+	efficiency = E / 6
```
Comment (tigercat2000) [inline code/modules/power/turbine.dm]: Won't this throw a division by 0 error if it has no manipulators in it for whatever reason?

Comment (Markolie) [inline code/modules/power/turbine.dm]: That won't ever happen unless an admin manually removes them from the component_parts list and then calls RefreshParts.

**Pull request merged**